File: app/src/main/java/io/goooler/demoapp/base/BaseApplication.java
package io.goooler.demoapp.base;

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.database.sqlite.SQLiteDatabase;
import android.os.Handler;

import com.squareup.leakcanary.LeakCanary;

import io.goooler.demoapp.model.Constants;

/**
 * 封装通用方法和一些初始化的动作
 */
public class BaseApplication extends Application {
    @SuppressLint("StaticFieldLeak")
    private static Context context;
    private static Handler handler;

    @Override
    public void onCreate() {
        super.onCreate();
        initGlobalObject();
        LeakCanary.install(this);
    }

    /**
     * 这里注意此方法仅在模拟器环境下会回调
     */
    @Override
    public void onTerminate() {
        super.onTerminate();
        destroyGlobalObject();
    }

    /**
     * 应用启动时初始化全局对象
     */
    private void initGlobalObject() {
        context = getApplicationContext();
        handler = new Handler();
    }

    /**
     * 应用结束时销毁全局对象
     * 真机环境需要在 mainActivity.onDestroy() 执行
     */
    public static void destroyGlobalObject() {
        handler.removeCallbacksAndMessages(null);
    }

    /**
     * 获取全局 context
     */
    public static Context getContext() {
        return context;
    }

    /**
     * 获取全局 handler
     */
    public static Handler getHandler() {
        return handler;
    }
}

File: app/src/main/java/io/goooler/demoapp/base/BaseApplication.kt
package io.goooler.demoapp.base

import android.annotation.SuppressLint
import android.content.Context
import androidx.multidex.MultiDexApplication
import com.alibaba.android.arouter.launcher.ARouter
import io.goooler.demoapp.util.CrashHandler
import io.goooler.demoapp.util.debugRun
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * 封装通用方法和一些初始化的动作
 */
class BaseApplication : MultiDexApplication() {

    override fun onCreate() {
        super.onCreate()
        initData()
    }

    /**
     * 应用启动时初始化
     */
    @SuppressLint("CheckResult")
    private fun initData() {
        context = applicationContext
        CrashHandler.init()
        ARouter.init(this)
        initLater()
    }

    private fun initLater() {
        GlobalScope.launch(Dispatchers.IO) {
            delay(2000)
            debugRun {
                ARouter.openLog()
                ARouter.openDebug()
            }
        }
    }

    companion object {
        /**
         * 获取全局 context
         */
        @SuppressLint("StaticFieldLeak")
        lateinit var context: Context
    }
}